### PR TITLE
Hide all packages by default

### PR DESCRIPTION
**Pull request by nickchapman-da**
Work to support issue #4094 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 1]
```diff
@@ -132,9 +132,9 @@ tests damlc repl davlDar = testGroup "Packaging"
             , "name: b"
             , "source: ."
             , "dependencies:"
-            , "  - daml-prim"
-            , "  - daml-stdlib"
             , "  - " <> aDar
+            , "build-options:"
+            , "- '--package=a-1.0'"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: This doesn’t seem right, why do I need to specify `a`? I explicitly depend on it in `build-options`.

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 2]
```diff
@@ -409,7 +409,10 @@ tests damlc repl davlDar = testGroup "Packaging"
             , "module B where"
             , "import A"
             ]
-          withCurrentDirectory projB $ callProcessSilent damlc ["build", "-o", "bar.dar"]
+          withCurrentDirectory projB $ callProcessSilent damlc
+            [ "build" , "-o", "bar.dar"
+            , "--package", "a-0.0.1"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: Same question here.

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 3]
```diff
@@ -429,7 +432,11 @@ tests damlc repl davlDar = testGroup "Packaging"
             , "import A"
             , "import B"
             ]
-          withCurrentDirectory projC $ callProcessSilent damlc ["build", "-o", "baz.dar"]
+          withCurrentDirectory projC $ callProcessSilent damlc
+            ["build" , "-o", "baz.dar"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: And here.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 4]
```diff
@@ -81,6 +81,20 @@ createProjectPackageDb opts thisSdkVer deps dataDeps = do
...
+
+    let versionize name = case name of
+          "daml-script" -> name <> "-" <> sdkVersion
+          "daml-trigger" -> name <> "-" <> sdkVersion
+          _ -> name
+
+    let units =
+          [ GHC.stringToUnitId (versionize base)
+          | dep <- deps
+          , let file = takeFileName dep
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: You can’t rely on the file names for figuring out the unit ids since these can be overwritten via `daml build -o`. We have a list of packages and their unit ids below at https://github.com/digital-asset/daml/blob/fcab7d1a0fa4c3ebd2a330519ebb57b1636382aa/compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs#L128 (note that this includes transitive dependencies so you don’t want all of those but you can use it to get the actual unit ids).

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 5]
```diff
@@ -81,6 +81,20 @@ createProjectPackageDb opts thisSdkVer deps dataDeps = do
...
     deps <- expandSdkPackages (filter (`notElem` basePackages) deps)
+
+    let versionize name = case name of
+          "daml-script" -> name <> "-" <> sdkVersion
+          "daml-trigger" -> name <> "-" <> sdkVersion
+          _ -> name
+
+    let units =
+          [ GHC.stringToUnitId (versionize base)
+          | dep <- deps
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: This also needs to take `dataDeps` into account.

File: compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Types.hs [hunk 6]
```diff
@@ -157,6 +170,7 @@ mkOptions opts@Options {..} = do
     ghcVersionFile <- locateRunfiles (mainWorkspace </> "compiler" </> "damlc" </> "ghcversion.h")
 
     pure opts {
+        optPackageImports = optPackageImports ++ basePackageImports optMbPackageName,
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Types.hs]: Instead of injecting the base imports here, I would suggest to do this in `initPackageDb` if `isProject` returns `False`.

File: compiler/damlc/lib/DA/Cli/Damlc.hs [hunk 7]
```diff
@@ -518,18 +520,27 @@ withPackageConfig projectPath f = do
...
+    then do
         isProject <- doesFileExist projectConfigName
-        when isProject $ do
+        if isProject then do
             withPackageConfig defaultProjectPath $ \PackageConfigFields {..} ->
                 createProjectPackageDb opts pSdkVersion pDependencies pDataDependencies
+        else
+           return []
+    else
+      return []
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs]: This breaks `--init-package-db=no`. We need to get the unit ids of the dependencies even if we are not reinitializing the package db. So the flag should probably be passed on to `createProjectPackageDb` which then always returns the unit ids but skips actually overwriting the package db. Given that we haven’t really advertised `--init-package-db=no` and #4413 should make it redundant, we might be able to just break it for now and handle it in a separate PR.

File: compiler/lsp-tests/src/Main.hs [hunk 8]
```diff
@@ -708,8 +708,10 @@ multiPackageTests damlc = testGroup "multi-package"
               Just escapedFpB <- pure $ escapeURIString isUnescapedInURIComponent <$> uriToFilePath (docB ^. uri)
               -- code lenses are a good test since they force LF compilation
               r <- getCodeLenses docB
+              -- TODO: test broken by --hide-all-packages being turned on. Why??
```
Comment (nickchapman-da) [inline compiler/lsp-tests/src/Main.hs]: I don't understand this!

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 1 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: this is sorted now

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 2 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: also sorted

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 3 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: same

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 4 above)
Comment (nickchapman-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: we don't use the names any more

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 5 above)
Comment (nickchapman-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: yes, it does now

File: compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Types.hs (hunk 6 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-opts/daml-opts-types/DA/Daml/Options/Types.hs]: The base import injection is now done in: 
`compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs`

File: compiler/damlc/lib/DA/Cli/Damlc.hs (hunk 7 above)
Comment (nickchapman-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs]: this issue still needs to be resolved...

File: compiler/lsp-tests/src/Main.hs (hunk 8 above)
Comment (nickchapman-da) [reply compiler/lsp-tests/src/Main.hs]: What I mean is. By turning on `--hide-all-packages`, this test broke...
So I made a temp hack to suppress the failure, showing what the new behaviour is.
But I have no idea if this new behaviour is correct. Or more generally what this test is about!

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs [hunk 9]
```diff
@@ -48,9 +49,22 @@ import DA.Daml.Preprocessor
...
+  | optMbPackageName == Just damlStdlib =
+      [ stringToUnitId "daml-prim" ]
+  | otherwise =
+      [ stringToUnitId "daml-prim"
+      , stringToUnitId damlStdlib ]
+
 -- | Convert to ghcide’s IdeOptions type.
 toCompileOpts :: Options -> Ghcide.IdeReportProgress -> Ghcide.IdeOptions
-toCompileOpts options@Options{..} reportProgress =
+toCompileOpts options@Options{..} reportProgress = do
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]:
```suggestion
toCompileOpts options@Options{..} reportProgress =
```
no need to make the diff unnecessarily noisy and this `do` doesn’t make any sense.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 10]
```diff
@@ -69,8 +69,34 @@ import SdkVersion
...
+            [ ( dropExtension $ takeFileName $ ZipArchive.eRelativePath e
+              , BSL.toStrict $ ZipArchive.fromEntry e
+              )
+            | e <- concatMap edMain dars
+            ]
+    dalfsFromFps <-
+        forM fpDalfs $ \fp -> do
+            bs <- BS.readFile fp
+            pure (dropExtension $ takeFileName fp, bs)
+    let allDalfs = dalfsFromDars ++ dalfsFromFps
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: This is a slightly confusing name especially given the similarity to https://github.com/digital-asset/daml/blob/36e188cac84ddb0f4259cd077a7bb6b4fd466311/compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs#L128. This does _not_ contain all dalfs, it only contains the main dalfs from each DAR. Maybe `rootDalfs` or just `mainDalfs` would be a better name.

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 11]
```diff
@@ -69,8 +69,34 @@ import SdkVersion
 --   ledger. Based on the DAML-LF we generate dummy interface files
 --   and then remap references to those dummy packages to the original DAML-LF
 --   package id.
-createProjectPackageDb :: NormalizedFilePath -> Options -> PackageSdkVersion -> [FilePath] -> [FilePath] -> IO ()
+
+calcUnitsFromDeps :: [FilePath] -> IO [UnitId]
+calcUnitsFromDeps deps = do
+    let (fpDars, fpDalfs) = partition ((== ".dar") . takeExtension) deps
+    dars <- mapM extractDar fpDars
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: This is a bit wasteful. We are now extracting each DAR twice, once here and then once at https://github.com/digital-asset/daml/blob/36e188cac84ddb0f4259cd077a7bb6b4fd466311/compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs#L115. We are also decoding everything twice, once here, and once at https://github.com/digital-asset/daml/blob/36e188cac84ddb0f4259cd077a7bb6b4fd466311/compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs#L130. Can we factor this such that we only have to do this once?

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 12]
```diff
@@ -132,8 +132,6 @@ tests damlc repl davlDar = testGroup "Packaging" $
             , "name: b"
             , "source: ."
             , "dependencies:"
-            , "  - daml-prim"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: Why did you change this?

File: compiler/lsp-tests/src/Main.hs (hunk 8 above)
Comment (cocreature) [reply compiler/lsp-tests/src/Main.hs]: This test checks that the IDE can load multiple packages at once. In this case the IDE will not be started in the project directory but in a parent directory, e.g., you open vscode in `foobar` and you have two projects at `foobar/a` and `foobar/b`. This means that `init` will do nothing since it is run in `foobar` and not in `foobar/a`. The proper fix for this is to change where we do `init` but I deliberately left this out of #4445 since it would have conflicted even more with this PR.

For now, I think the best fix is to add a `--package` flag to the `build-options` field of `b` and mention this in the docs at https://github.com/digital-asset/daml/blob/master/docs/source/daml/daml-studio.rst#working-with-multiple-packages. We can then shuffle around pkg db initialization afterwards to make this unnecessary.

Comment (cocreature) [reply compiler/lsp-tests/src/Main.hs]: https://github.com/digital-asset/daml/pull/4481 adds another test for starting the IDE in the respective project directory. In that case `init` should work correctly and you should not need to add a `--package` flag. Probably a good idea to merge #4481 first and rebase this on top of it to make sure this is the case.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs (hunk 9 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: done
(there was `let` added here, but it got moved in `getDamlGhcSession` after the last rebase)

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 10 above)
Comment (nickchapman-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: done

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 11 above)
Comment (nickchapman-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: as discussed.
our new plan to relocate the `calUnitsForDeps` code means this sharing wont be possible now

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 12 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: :man_shrugging: 
I put it back

File: compiler/damlc/lib/DA/Cli/Damlc.hs (hunk 7 above)
Comment (nickchapman-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs]: The changes were removed from `Damlc.hs`, so nothing to do here

File: compiler/lsp-tests/src/Main.hs (hunk 8 above)
Comment (nickchapman-da) [reply compiler/lsp-tests/src/Main.hs]: These tests all work now.

File: compiler/damlc/daml-opts/BUILD.bazel [hunk 13]
```diff
@@ -44,14 +45,22 @@ da_haskell_library(
         "ghc-lib-parser",
         "ghcide",
         "mtl",
-        "text",
+        "safe-exceptions",
         "shake",
+        "text",
+        "transformers",
+        "utf8-string",
+        "zip",
```
Comment (cocreature) [inline compiler/damlc/daml-opts/BUILD.bazel]: Why do we need `zip`? I only see an import of `zip-archive`.

File: compiler/damlc/daml-opts/BUILD.bazel [hunk 14]
```diff
@@ -44,14 +45,22 @@ da_haskell_library(
         "ghc-lib-parser",
         "ghcide",
         "mtl",
-        "text",
+        "safe-exceptions",
         "shake",
+        "text",
+        "transformers",
+        "utf8-string",
```
Comment (cocreature) [inline compiler/damlc/daml-opts/BUILD.bazel]: I don’t see an import of `utf8-string`.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs [hunk 15]
```diff
@@ -83,10 +96,18 @@ getDamlGhcSession :: Options -> IO (FilePath -> Action HscEnvEq)
 getDamlGhcSession options@Options{..} = do
     findProjectRoot <- memoIO findProjectRoot
     getSession <- memoIO $ \mbProjectRoot -> do
+        let base = mkBaseUnits optMbPackageName
+        optPackageImports <-
+          if getInferDependantPackages optInferDependantPackages
+          then do
+            more <- dependantUnitsFromDamlYaml mbProjectRoot
+            pure $ optPackageImports ++ map mkPackageFlag (base ++ more)
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: I think it probably makes more sense to put `optPackageImports` at the end so you can infer the inferred flags.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs [hunk 16]
```diff
@@ -399,3 +419,90 @@ checkDFlags Options {..} dflags@DynFlags {..}
...
+    expand mbSdkPath fp
+      | fp `elem` basePackages = pure fp
+      | isSdkPackage fp = case mbSdkPath of
+            Just sdkPath -> pure $ sdkPath </> "daml-libs" </> fp <.> "dar"
+            Nothing -> fail $ "Cannot resolve SDK dependency '" ++ fp ++ "'. Use daml assistant."
+      | otherwise = pure fp
+
+
+mkPackageFlag :: UnitId -> PackageFlag
+mkPackageFlag unitId = ExposePackage "--package " (UnitIdArg unitId) (ModRenaming True [])
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: The string here should include the unit id. See https://github.com/digital-asset/daml/pull/4516.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs [hunk 17]
```diff
@@ -399,3 +419,90 @@ checkDFlags Options {..} dflags@DynFlags {..}
...
+  | optMbPackageName == Just damlStdlib =
+      [ stringToUnitId "daml-prim" ]
+  | otherwise =
+      [ stringToUnitId "daml-prim"
+      , stringToUnitId damlStdlib ]
+
+dependantUnitsFromDamlYaml :: Maybe FilePath -> IO [UnitId]
+dependantUnitsFromDamlYaml mbRoot = do
+  let dir = fromMaybe "." mbRoot
+  withCurrentDirectory dir $ do
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: Why do we have to change the directory here? Is that only so that `calcUnitsFromDeps` resolves relative file paths directly? If so, I would prefer to instead pass on the directory to `calcUnitsFromDeps`. CWD is process-global state so modifying it is a bit fishy and will almost certainly break in some cases without a lock.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs [hunk 18]
```diff
@@ -399,3 +419,90 @@ checkDFlags Options {..} dflags@DynFlags {..}
...
+  dalfsFromFps <-
+    forM fpDalfs $ \fp -> do
+      bs <- BS.readFile fp
+      pure (dropExtension $ takeFileName fp, bs)
+  let mainDalfs = dalfsFromDars ++ dalfsFromFps
+  flip mapMaybeM mainDalfs $ \(name, dalf) -> runMaybeT $ do
+    (pkgId, _package) <-
+        liftIO $
+        either (fail . DA.Pretty.renderPretty) pure $
+        Archive.decodeArchive Archive.DecodeAsMain dalf
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]:
```suggestion
        Archive.decodeArchivePackageId dalf
```
See https://github.com/digital-asset/daml/pull/4550

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs [hunk 19]
```diff
@@ -399,3 +419,90 @@ checkDFlags Options {..} dflags@DynFlags {..}
...
+        [ ( dropExtension $ takeFileName $ ZipArchive.eRelativePath e
+          , BSL.toStrict $ ZipArchive.fromEntry e)
+        | e <- entries ]
+  dalfsFromFps <-
+    forM fpDalfs $ \fp -> do
+      bs <- BS.readFile fp
+      pure (dropExtension $ takeFileName fp, bs)
+  let mainDalfs = dalfsFromDars ++ dalfsFromFps
+  flip mapMaybeM mainDalfs $ \(name, dalf) -> runMaybeT $ do
+    (pkgId, _package) <-
```
Comment (cocreature) [inline compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]:
```suggestion
    pkgId <-
```

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 20]
```diff
@@ -272,14 +272,14 @@ generateAndInstallIfaceFiles dalf src opts workDir dbPath projectPackageDatabase
...
             , optMbPackageName = Just unitIdStr
-            , optHideAllPkgs = True
             , optGhcCustomOpts = []
             , optPackageImports =
                   baseImports ++
                   depImps ++
                   [ exposePackage (GHC.stringToUnitId $ takeBaseName dep) True []
                   | dep <- deps
                   ]
+            , optInferDependantPackages = InferDependantPackages False
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: Maybe add a comment here “When compiling dummy interface files for a data-dependency we know all package flags so we don’t need to infer anything.”.

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 21]
```diff
@@ -134,7 +134,6 @@ tests damlc repl davlDar oldProjDar = testGroup "Packaging" $
             , "source: ."
             , "dependencies:"
             , "  - daml-prim"
-            , "  - daml-stdlib"
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: You still removed `daml-stdlib` here which should be able to stay afaict.

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 22]
```diff
@@ -1186,7 +1164,12 @@ dataDependencyTests damlc repl davlDar oldProjDar = testGroup "Data Dependencies
 
 -- | Only displays stdout and stderr on errors
 callProcessSilent :: FilePath -> [String] -> IO ()
-callProcessSilent cmd args = do
+callProcessSilent prog args = do
+  putStrLn $ "PP: " <> show (prog,args)
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: Looks like leftover debugging output.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs (hunk 15 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: I did as you suggest. I assume this is because later options can override/extend earlier ones.

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs (hunk 17 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: ok, `withCurrentDirectory` is removed

File: compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs (hunk 18 above)
Comment (nickchapman-da) [reply compiler/damlc/daml-opts/daml-opts/DA/Daml/Options.hs]: rebased, and made this change

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 22 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: Aggh! Thank you

**Comment by nickchapman-da** [comment]
All comments addressed.
